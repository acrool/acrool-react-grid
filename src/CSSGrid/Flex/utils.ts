import {
    TStyledProps,
    IGridProps,
    TGridGap,
    TGridTemplate,
    TGridGaps,
    TGridCol,
    TSpan,
    IGridColProps, TFlexDirection, IFlexProps, TDirection
} from '../../types';
import {noXsMediaSizes} from '../../config';
import media from '../../media';
import {css} from 'styled-components';





interface ICSSGetter {
    direction: (direction: TDirection) => string

    gap: (gapNum: TGridGap) => string
    columnGap: (gapNum: TGridGap) => string
    rowGap: (gapNum: TGridGap) => string
}

export const cssGetter: ICSSGetter = {
    direction: (direction) => {
        return `flex-direction: ${direction};`;
    },
    gap: (gapNum) => {
        return `
        gap: ${gapNum};
        `;
    },
    columnGap: (gapNum) => {
        return `
        column-gap: ${gapNum};
        `;
    },
    rowGap: (gapNum) => {
        return `
        row-gap: ${gapNum};
        `;
    },
};



/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IFlexProps>) => {
    return noXsMediaSizes
        .map(sizeName => {
            const mediaColSpanProps = typeof props.direction === 'object' ? props.direction[sizeName]: undefined;
            const mediaGapProps = typeof props.gap === 'object' ? props.gap[sizeName]: undefined;
            const mediaColumnGapProps = typeof props.columnsGap === 'object' ? props.columnsGap[sizeName]: undefined;
            const mediaRowGapProps = typeof props.rowsGap === 'object' ? props.rowsGap[sizeName]: undefined;

            return media[sizeName]`

            ${typeof mediaColSpanProps !== 'undefined' && css`
                ${cssGetter.direction(mediaColSpanProps)};
            `}

            ${typeof mediaGapProps !== 'undefined' && css`
                ${cssGetter.gap(mediaGapProps)};
            `}
            ${typeof mediaColumnGapProps !== 'undefined' && css`
                ${cssGetter.columnGap(mediaColumnGapProps)};
            `}
            ${typeof mediaRowGapProps !== 'undefined' && css`
                ${cssGetter.rowGap(mediaRowGapProps)};
            `}
    `;
        });
};

