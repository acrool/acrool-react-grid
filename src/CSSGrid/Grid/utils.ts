import {TStyledProps, IGridProps, TGridGap, TGridTemplate, TGridGaps} from '../../types';
import {noXsMediaSizes} from '../../config';
import media from '../../media';
import {css} from 'styled-components';




interface ICSSGetterCss {
    columns: (columns: TGridTemplate) => string
    rows: (rows: TGridTemplate) => string
    gap: (gapNum: TGridGaps) => string
    columnGap: (gapNum: TGridGap) => string
    rowGap: (gapNum: TGridGap) => string
}


export const cssGetter: ICSSGetterCss = {
    columns: (columns) => {

        if(typeof columns === 'string'){
            return `
            grid-template-columns: ${columns};
            `;
        }
        return `
            grid-template-columns: repeat(${columns}, auto)
        `;
    },
    rows: (rows) => {
        if(typeof rows === 'string'){
            return `
            grid-template-rows: ${rows};
            `;
        }
        return `
            grid-template-rows: ${rows === 1 ? '1fr' : `repeat(${rows}, 1fr)`}
        `;
    },
    gap: (gapNum) => {
        return `
        gap: ${gapNum};
        `;
    },
    columnGap: (gapNum) => {
        return `
        column-gap: ${gapNum};
        `;
    },
    rowGap: (gapNum) => {
        return `
        row-gap: ${gapNum};
        `;
    },
};



/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IGridProps>) => {
    return noXsMediaSizes
        .map(sizeName => {
            const mediaColumnProps = typeof props.columns === 'object' ? props.columns[sizeName]: undefined;
            const mediaRowProps = typeof props.rows === 'object' ? props.rows[sizeName]: undefined;
            return media[sizeName]`
            ${typeof mediaColumnProps !== 'undefined' && css`
                ${cssGetter.columns(mediaColumnProps)};
            `}
            ${typeof mediaRowProps !== 'undefined' && css`
                ${cssGetter.rows(mediaRowProps)};
            `}
        `;
        });
};

