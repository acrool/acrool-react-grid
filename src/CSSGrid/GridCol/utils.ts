import {
    TStyledProps,
    IGridProps,
    TGridGap,
    TGridTemplate,
    TGridGaps,
    TGridCol,
    TSpan,
    IGridColProps
} from '../../types';
import {noXsMediaSizes} from '../../config';
import media from '../../media';
import {css} from 'styled-components';





interface ICSSGetter {
    col: (span: TSpan) => string
    row: (span: TSpan) => string
}

export const cssGetter: ICSSGetter = {
    col: (colSpan) => {
        return `grid-column: auto/span ${colSpan};`;
    },
    row: (rowSpan) => {
        return `grid-row: auto/span ${rowSpan};`;
    },
};



/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IGridColProps>) => {
    return noXsMediaSizes
        .map(sizeName => {
            const mediaColSpanProps = typeof props.col === 'object' ? props.col[sizeName]: undefined;
            const mediaRowSpanProps = typeof props.row === 'object' ? props.row[sizeName]: undefined;

            return media[sizeName]`

            ${typeof mediaColSpanProps !== 'undefined' && css`
                ${cssGetter.col(mediaColSpanProps)};
            `}
            ${typeof mediaRowSpanProps !== 'undefined' && css`
                ${cssGetter.row(mediaRowSpanProps)};
            `}
    `;
        });
};

