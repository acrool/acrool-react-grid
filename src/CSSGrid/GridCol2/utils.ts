import {
    TStyledProps,
    IGridProps,
    TGridGap,
    TGridTemplate,
    TGridGaps,
    TGridCol,
    TSpan,
    IGridColProps
} from '../../types';
import {noXsMediaSizes} from '../../config';
import media from '../../media';
import {css} from 'styled-components';





interface ICSSGetter {
    col: (column: TGridCol) => string
    colSpans: (span: TSpan) => string
    rowSpans: (span: TSpan) => string
}

export const cssGetter: ICSSGetter = {
    col: (col) => {
        return `grid-column: auto/span ${col};`;
    },
    colSpans: (colSpan) => {
        return `
            grid-column-start: span ${colSpan};
        `;
    },
    rowSpans: (rowSpan) => {
        return `
            grid-row-start: span ${rowSpan};
        `;
    }
};



/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IGridColProps>) => {
    return noXsMediaSizes
        .map(sizeName => {
            const mediaColSpanProps = typeof props.colSpans === 'object' ? props.colSpans[sizeName]: undefined;
            const mediaRowSpanProps = typeof props.rowSpans === 'object' ? props.rowSpans[sizeName]: undefined;

            const mediaArg = props[sizeName];
            return media[sizeName]`
            ${typeof mediaArg !== 'undefined' && css`
                ${cssGetter.col(mediaArg)};
            `}

            ${typeof mediaColSpanProps !== 'undefined' && css`
                ${cssGetter.colSpans(mediaColSpanProps)};
            `}
            ${typeof mediaRowSpanProps !== 'undefined' && css`
                ${cssGetter.rowSpans(mediaRowSpanProps)};
            `}
    `;
        });
};

