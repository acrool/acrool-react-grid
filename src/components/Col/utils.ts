import {TStyledProps, IColProps} from '../../types';
import {mediaSizes, themeName} from '../../config';
import media from '../../media';
import {css} from 'styled-components';
import getCss from './css';

/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IColProps>) => {
    return mediaSizes
        .map(sizeName => {
            return media[sizeName]`
            ${typeof props[sizeName] !== 'undefined' && css`
                ${getCss.col(props[sizeName], props.theme[themeName]?.gridColumns)};
            `}
    `;
        });
};
