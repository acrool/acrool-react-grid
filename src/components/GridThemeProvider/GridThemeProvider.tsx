import React, {PropsWithChildren} from 'react';
import {ThemeProvider, createGlobalStyle} from 'styled-components';

import {defaultGridTheme, themeName} from '../../config';
import {IGridSetting} from '../../types';


interface IProps {
    gridTheme: Partial<IGridSetting>;
    children: React.ReactNode;
}

const ThemeProviderFixed = ThemeProvider as unknown as React.FC<PropsWithChildren<{ theme: any }>>;


/**
 * Grid Theme Provider
 * use :
 *     <GridThemeProvider gridTheme={{
 *         gridGutterWidth: 5,
 *     }}>
 *          <Container>, <Row>, <Col> ...then eat this setting
 *     </GridThemeProvider>
 */
const GridThemeProvider = (props: IProps) => {
    const {gridTheme, children} = props;
    const {gridBreakpoints, containerMaxWidths, gridGutterWidthMedia, ...baseDefaultTheme} = defaultGridTheme;

    // console.log('test render bear-react-grid version 10');
    const composeGridTheme = {
        [themeName]: {
            ...baseDefaultTheme,
            ...gridTheme,
            gridBreakpoints: Object.assign({}, gridBreakpoints, gridTheme?.gridBreakpoints),
            containerMaxWidths: Object.assign({}, containerMaxWidths, gridTheme?.containerMaxWidths),
            gridGutterWidthMedia: Object.assign({}, gridGutterWidthMedia, gridTheme?.gridGutterWidthMedia),
        }
    };

    return <ThemeProviderFixed theme={composeGridTheme}>
        {children}
        <GlobalStyled/>
    </ThemeProviderFixed>;
};

export default GridThemeProvider;


const GlobalStyled = createGlobalStyle`
    
`;
