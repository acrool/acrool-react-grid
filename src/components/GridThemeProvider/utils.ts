import {TStyledProps, IRowProps, NoXsMediaSize} from '../../types';
import {mediaSizes} from '../../config';
import media from '../../media';


const renderMarginStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .m-${sizeName}-${idx} {margin: var(--bear-spacer-${idx}) !important;}
        .mt-${sizeName}-${idx}, .my-${sizeName}-${idx} {margin-top: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .mr-${sizeName}-${idx}, .mx-${sizeName}-${idx} {margin-right: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .mb-${sizeName}-${idx}, .my-${sizeName}-${idx} {margin-bottom: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .ml-${sizeName}-${idx}, .mx-${sizeName}-${idx} {margin-left: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        `;
    }).join('');
};

const renderPaddingStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .p-${sizeName}-${idx} {margin: var(--bear-spacer-${idx}) !important;}
        .pt-${sizeName}-${idx}, .py-${sizeName}-${idx} {padding-top: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pr-${sizeName}-${idx}, .px-${sizeName}-${idx} {padding-right: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pb-${sizeName}-${idx}, .py-${sizeName}-${idx} {padding-bottom: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pl-${sizeName}-${idx}, .px-${sizeName}-${idx} {padding-left: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        `;
    }).join('');
};

const renderOrderStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .order-${sizeName}-${idx} {order: ${idx}!important;}
        `;
    }).join('');
};


/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IRowProps>) => {
    return mediaSizes
        .map(sizeName => {
            return media[sizeName]`
            .d-${sizeName}-none {display: none !important;}
            .d-${sizeName}-inline {display: inline !important;}
            .d-${sizeName}-inline-block {display: inline-block !important;}
            .d-${sizeName}-block {display: block !important;}
            .d-${sizeName}-flex {display: flex !important;}
            .d-${sizeName}-inline-flex {display: inline-flex !important;}
            .d-${sizeName}-table-cell {display: table-cell !important;}
            .d-${sizeName}-table {display: table !important;}
            .d-${sizeName}-table-row {display: table-row !important;}

            .flex-${sizeName}-row {flex-direction: row !important;}
            .flex-${sizeName}-column {flex-direction: column !important;}
            .flex-${sizeName}-row-reverse {flex-direction: row-reverse !important;}
            .flex-${sizeName}-column-reverse {flex-direction: column-reverse !important;}

            .flex-${sizeName}-wrap {flex-wrap: wrap !important;}
            .flex-${sizeName}-nowrap {flex-nowrap: nowrap !important;}
            .flex-${sizeName}-wrap-reverse {flex-nowrap: wrap-reverse !important;}

            .flex-${sizeName}-fill {flex: 1 1 auto !important;}
            .flex-${sizeName}-grow-0 {flex-grow: 0 !important;}
            .flex-${sizeName}-grow-1 {flex-grow: 1 !important;}
            .flex-${sizeName}-shrink-0 {flex-shrink: 0 !important;}
            .flex-${sizeName}-shrink-1 {flex-shrink: 1 !important;}

            .justify-content-${sizeName}-start {justify-content: flex-start !important;}
            .justify-content-${sizeName}-end {justify-content: flex-end !important;}
            .justify-content-${sizeName}-center {justify-content: flex-center !important;}
            .justify-content-${sizeName}-between {justify-content: space-between !important;}
            .justify-content-${sizeName}-around {justify-content: space-around !important;}

            .align-items-${sizeName}-start {align-items: flex-start !important;}
            .align-items-${sizeName}-end {align-items: flex-end !important;}
            .align-items-${sizeName}-center {align-items: center !important;}
            .align-items-${sizeName}-baseline {align-items: baseline !important;}
            .align-items-${sizeName}-stretch {align-items: stretch !important;}

            .align-content-${sizeName}-start {align-content: start !important;}
            .align-content-${sizeName}-end {align-content: flex-end !important;}
            .align-content-${sizeName}-center {align-content: center !important;}
            .align-content-${sizeName}-between {align-content: between !important;}
            .align-content-${sizeName}-around {align-content: around !important;}
            .align-content-${sizeName}-stretch {align-content: stretch !important;}

            .align-self-${sizeName}-start {align-self: flex-start !important;}
            .align-self-${sizeName}-end {align-self: flex-end !important;}
            .align-self-${sizeName}-center {align-self: center !important;}
            .align-self-${sizeName}-baseline {align-self: baseline !important;}
            .align-self-${sizeName}-stretch {align-self: stretch !important;}

            .text-${sizeName}-justify {text-align: justify !important;}
            .text-${sizeName}-left {text-align: left !important;}
            .text-${sizeName}-right {text-align: right !important;}
            .text-${sizeName}-center {text-align: center !important;}

            .m-${sizeName}-auto {margin: var(--bear-spacer-auto) !important;}
            .mt-${sizeName}-auto, .my-${sizeName}-auto {margin-top: auto !important;}
            .mr-${sizeName}-auto, .mx-${sizeName}-auto {margin-right: auto !important;}
            .mb-${sizeName}-auto, .my-${sizeName}-auto {margin-bottom: auto !important;}
            .ml-${sizeName}-auto, .mx-${sizeName}-auto {margin-left: auto !important;}

            ${renderMarginStyle(sizeName)}
            ${renderPaddingStyle(sizeName)}
            ${renderOrderStyle(sizeName)}

            .overflow-x-${sizeName}-hidden {overflow-x: hidden!important;}
            .overflow-y-${sizeName}-hidden {overflow-y: hidden!important;}
            .overflow-x-${sizeName}-auto {overflow-x: auto!important;}
            .overflow-y-${sizeName}-auto {overflow-y: auto!important;}
        `;
        });
};

