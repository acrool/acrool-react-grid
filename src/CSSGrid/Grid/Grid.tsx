import React from 'react';
import styled, {css} from 'styled-components';
import {TStyledProps, IGridProps} from '../../types';
import {themeName} from '../../config';
import {createBreakpoint, cssGetter, createInfo} from './utils';




/**
 * Grid Component
 * 由外部設定柵欄顯示
 */
const Grid = styled.div.attrs((props: TStyledProps<IGridProps>) => ({
    'data-grid': createInfo(props),
}))`
    --bear-gutter-x: ${props => props.theme[themeName]?.gutter};
    --bear-gutter-y: 0;

    display: grid;
    grid-template-rows: auto;
    grid-template-columns: repeat(${props => props.theme[themeName]?.gridColumns}, 1fr);
    column-gap: var(--bear-gutter-x);
    row-gap: var(--bear-gutter-y);


    ${(props: TStyledProps<IGridProps>) => css`
     ${props.col && cssGetter.col(props.col)};

     ${createBreakpoint(props)};
   `}

`;

export default Grid;

