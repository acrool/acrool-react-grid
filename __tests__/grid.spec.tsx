import React from 'react';
import 'jest-styled-components';
import {render} from '@testing-library/react';
import '@testing-library/dom';
import GridThemeProvider from '../src/GridThemeProvider';
import Grid from '../src/CSSGrid/Grid';
import {auto} from '../src/utils';



describe('Grid', () => {
    it('renders grid default', () => {
        const {container} = render(<Grid>col</Grid>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders col-auto', () => {
        const {container} = render(<Grid col={auto(1)}>
            <div>col 1</div>
            <div>col 2</div>
        </Grid>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders col-6', () => {
        const {container} = render(<Grid col={6}>
            <div>col 1</div>
            <div>col 2</div>
            <div>col 3</div>
            <div>col 4</div>
            <div>col 5</div>
            <div>col 6</div>
        </Grid>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });
});
