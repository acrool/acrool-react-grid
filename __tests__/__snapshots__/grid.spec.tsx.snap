// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders col-6 1`] = `
.c0 {
  --acrool-gutter-x: 1.5rem;
  --acrool-gutter-y: 0;
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: repeat(12,1fr);
  -webkit-column-gap: var(--acrool-gutter-x);
  column-gap: var(--acrool-gutter-x);
  row-gap: var(--acrool-gutter-y);
  grid-template-columns: repeat(6,1fr);
}

<div
  class="c0"
  data-grid="grid-6"
>
  <div>
    col 1
  </div>
  <div>
    col 2
  </div>
  <div>
    col 3
  </div>
  <div>
    col 4
  </div>
  <div>
    col 5
  </div>
  <div>
    col 6
  </div>
</div>
`;

exports[`Grid renders col-auto 1`] = `
.c0 {
  --acrool-gutter-x: 1.5rem;
  --acrool-gutter-y: 0;
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: repeat(12,1fr);
  -webkit-column-gap: var(--acrool-gutter-x);
  column-gap: var(--acrool-gutter-x);
  row-gap: var(--acrool-gutter-y);
  grid-template-columns: repeat(1,auto);
}

<div
  class="c0"
  data-grid="grid-[repeat(1, auto)]"
>
  <div>
    col 1
  </div>
  <div>
    col 2
  </div>
</div>
`;

exports[`Grid renders grid default 1`] = `
.c0 {
  --acrool-gutter-x: 1.5rem;
  --acrool-gutter-y: 0;
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: repeat(12,1fr);
  -webkit-column-gap: var(--acrool-gutter-x);
  column-gap: var(--acrool-gutter-x);
  row-gap: var(--acrool-gutter-y);
}

<div
  class="c0"
  data-grid=""
>
  col
</div>
`;
